streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java

```diff
@@ -0,0 +1,170 @@
+            }
+        }
+
+        // Then process any error tasks that are ready to be retried
+        for (final ErrorTask task : errorTasks.values()) {
+            if (task.readyForRetry()) {
+                try {
+                    totalProcessed += processTask(task.streamTask, maxNumRecords, time);
+                    log.info("Removing task {} from penalty box after clean run", task.streamTask.id());
+                    tasks.topologyMetadata().clearErrorStatusForTask(task.streamTask.id());
+                } catch (final Throwable throwable) {
+                    log.debug("Keeping task {} to penalty box due to hitting error", task.streamTask.id());
+                }
+            }
+        }
+
+        return totalProcessed;
+    }
+
+    private long processTask(final Task task, final int maxNumRecords, final Time time) {
+        int processed = 0;
+        long now = time.milliseconds();
+
+        final long then = now;
+        try {
+            while (processed < maxNumRecords && task.process(now)) {
+                task.clearTaskTimeout();
+                processed++;
+            }
+        } catch (final TimeoutException timeoutException) {
+            task.maybeInitTaskTimeoutOrThrow(now, timeoutException);
+            log.debug(
+                String.format(
+                    "Could not complete processing records for %s due to the following exception; will move to next task and retry later",
+                    task.id()),
+                timeoutException
+            );
+        } catch (final TaskMigratedException e) {
+            log.info("Failed to process stream task {} since it got migrated to another thread already. " +
+                "Will trigger a new rebalance and close all tasks as zombies together.", task.id());
+            throw e;
+        } catch (final StreamsException e) {
+            log.error("Failed to process stream task {} due to the following error:", task.id(), e);
+            e.setTaskId(task.id());
+            throw e;
+        } catch (final RuntimeException e) {
+            log.error("Failed to process stream task {} due to the following error:", task.id(), e);
+            throw new StreamsException(e, task.id());
+        } finally {
+            now = time.milliseconds();
+            task.recordProcessBatchTime(now - then);
+        }
+        return processed;
+    }
+
+    public boolean isInError(final TaskId taskId) {
+        return errorTasks.containsKey(taskId);
+    }
+
+    static class ErrorTask implements Comparable<ErrorTask> {
+
+        private final StreamTask streamTask;
+        private Queue<Throwable> errors = new LinkedList<>();
+
+        public ErrorTask(final StreamTask streamTask) {
+            this.streamTask = streamTask;
+        }
+
+        public boolean readyForRetry() {
+            // TODO: implement backoff for persisting errors
+            return true;
+        }
+
+        public void addError(final Throwable throwable) {
```

Comment:
The following functions are not used in this PR, assuming they will be used in future PRs.